src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -234,8 +214,7 @@ TryDropOrphanedResources(bool waitForLocks)
 
 	PG_TRY();
 	{
-		droppedResourceCount = DropOrphanedShardsForMove(waitForLocks);
-		droppedResourceCount += DropOrphanedResourcesForCleanup();
+		droppedResourceCount = DropOrphanedResourcesForCleanup();
```

Comment:
I didn't realise that we don't have waitForLocks for the new function. Any clue why the old one had this parameter? Should we have it too for the new function?

Reply:
>  Before doing any of this it will take an exclusive PlacementCleanup lock.
 This is to ensure that this function is not being run concurrently.
 Otherwise really bad race conditions are possible, such as removing all
 placements of a shard. waitForLocks indicates if this function should
 wait for this lock or not.

Seems like it was to prevent dropping the placements that shouldn't be dropped. But with the `pg_dist_cleanup` records, we have the node group and object name so it doesn't look like it would cause such an issue. Thoughts?

Reply:
We now (always) lock at the operation ID level, where we previously had a global lock.

One high-level confusing thing is that if 2 processes run cleanup at the same time, they will race for operation IDs. As in, they might both get some arbitrary subset of operation IDs, but they will be distinct subsets.